### PR TITLE
Clarify error for non-UTF-8 encoded data in runpath setup

### LINKED ISSUE #6951: Improve error message with the names of files with non utf-8 characters
**Describe the bug**
Having non "utf-8" characters in input files (in the reported case it was DATA-file for Eclipse, and  ERT template file) gives an error, but does not print the name of the files that caused it 

**To reproduce**
Reported by Therese Natterøy on behalf of another user, no steps to reproduce provided.

**Expected behaviour**
A error message contain the name of problematic files.

**Environment**
 `/prog/res/komodo/2023.11.06-py38-rhel7`

### PR DESCRIPTION
**Pull request by dafeda**
**Issue**
Resolves #6951

**Approach**
_Short description of the approach_

(Screenshot of new behavior in GUI if applicable)

- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: docs/reference/configuration/keywords.rst [hunk 1]
```diff
@@ -128,44 +128,27 @@ Commonly used keywords
...
-        This is used to replace ERT magic strings into the data file, as well as
-        update the number of cpus that are reserved for ERT in the queue system.
-
-        It searches for PARALLEL in the data file, and if that is not found it
-        will search for SLAVE and update <NUM_CPU> according to how many nodes are
-        found, note that it does *not* parse the data files of the nodes, and will
-        assume one cpu per node where entry number 5 is not set, and the number of
-        entry number 5 otherwise plus one cpu for the master node.
+        - ert will search for ``PARALLEL`` in the data file, and it not found, it will search for ``SLAVE`` and update :ref:`NUM_CPU <num_cpu>` according to how many nodes are found.
+        - Note that it does *not* parse the data files of the nodes, and will assume one cpu per node where entry number 5 is not set, and the number of entry number 5 otherwise plus one cpu for the master node.
```
Comment (dafeda) [inline docs/reference/configuration/keywords.rst:141]: I don't understand this. What is entry number 5?

File: docs/reference/configuration/keywords.rst [hunk 2]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
 .. _data_file:
 .. topic:: DATA_FILE
 
-        Meant to be set to the filepath of an eclipse simulator input, when such
-        a simulator is used. This does two things. First, the DATA_FILE will be
-        templated, see :ref:`RUN_TEMPLATE <run_template>`. Second, ert will look
-        for the PARALLEL keyword in this file in order to set :ref:`NUM_CPU <num_cpu>`.
+    Specify the filepath to the ``.DATA`` file of Eclipse/flow.
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:131]: nit-picking: I believe `ECLIPSE` is usually capitalised and I would also use `OPM Flow` instead of flow

File: docs/reference/configuration/keywords.rst [hunk 3]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
...
+    This does two things:
 
-        The templated file will be named according to :ref:`ECLBASE <ECLBASE>`
-        and copied to the runpath folder. Note that support for parsing the
-        ECLIPSE data file is limited, and using explicit templating with
-        :ref:`RUN_TEMPLATE <run_template>` is recommended where possible.
+    1. Template the ``DATA_FILE`` using :ref:`RUN_TEMPLATE <run_template>`.
 
+        The templated file will be named according to :ref:`ECLBASE <ECLBASE>` and
+        copied to the runpath folder. Note that support for parsing the Eclipse/flow
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:137]: and here too

File: docs/reference/configuration/keywords.rst (hunk 2 above)
Comment (dafeda) [reply docs/reference/configuration/keywords.rst:131]: Would you mind using the `Suggest` feature in GitHub so that I can just commit your proposed changes directly from GitHub?

File: tests/unit_tests/test_run_path_creation.py [hunk 4]
```diff
@@ -194,6 +194,33 @@ def test_run_template_replace_in_ecl_data_file(key, expected, prior_ensemble):
...
+        """
+    NUM_REALIZATIONS 1
+    ECLBASE ECL_CASE<IENS>
+    DATA_FILE MY_DATA_FILE.DATA
+    """
+    )
+    Path("MY_DATA_FILE.DATA").write_text("I WANT TO REPLACE:<DATE>", encoding="utf-8")
+    Path("config.ert").write_text(config_text, encoding="utf-8")
+
+    ert_config = ErtConfig.from_file("config.ert")
```
Comment (oyvindeide) [inline tests/unit_tests/test_run_path_creation.py:208]: This looks like a good improvement, but as we already know about this file when ert starts, perhaps we should validate it there? Then the user would get a message in the Suggester instead of when they start the model. We could of course do both in case they change it after ert has started.

File: docs/reference/configuration/keywords.rst [hunk 5]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
...
 
-        It searches for PARALLEL in the data file, and if that is not found it
-        will search for SLAVE and update <NUM_CPU> according to how many nodes are
-        found, note that it does *not* parse the data files of the nodes, and will
-        assume one cpu per node where entry number 5 is not set, and the number of
-        entry number 5 otherwise plus one cpu for the master node.
+        ert will search for ``PARALLEL`` in the data file, and if not found, will
+        search for ``SLAVES`` and update :ref:`NUM_CPU <num_cpu>` according to how
+        many nodes are found.
+        The fifth entry of the keyword ``SLAVES`` determines the number of processors
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:146]: This is a bad wording to me. The keyword ``SLAVES`` is defined through a set of up to 5 numbers:
```
SLAVES
--    slave   datafile    machine        directory                    number of
--    name    root        hostname      of data file                 processors
--
      'upper'  'base'       '*'        '/usr/models/upper'            1    /
      'lower'  'base'       '*'        '/usr/models/lower'            5   /

/

```

Comment (dafeda) [reply docs/reference/configuration/keywords.rst:146]: What do you suggest?

Comment (kvashchuka) [reply docs/reference/configuration/keywords.rst:146]: and I read `fifth entry of the keyword ``SLAVES`` ` as if this keyword will be entered 5 times. But I am also not familiar with the vocabulary, so I might be wrong

File: tests/unit_tests/test_run_path_creation.py (hunk 4 above)
Comment (dafeda) [reply tests/unit_tests/test_run_path_creation.py:208]: I think we already do that.

See this test: https://github.com/equinor/ert/blob/5072a4eb4c17a20d45efa81df191c5728518d5fe/tests/unit_tests/config/test_ert_config.py#L423-L451

Are you perhaps thinking of something else?

File: docs/reference/configuration/keywords.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (dafeda) [reply docs/reference/configuration/keywords.rst:146]: I think of mambo number five when I read that.

File: docs/reference/configuration/keywords.rst [hunk 6]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
...
+        to run the slave on.
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:147]:
```suggestion
        to run the slave reservoir on.
```

File: docs/reference/configuration/keywords.rst (hunk 2 above)
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:131]:
```suggestion
    Specify the filepath to the ``.DATA`` file of ECLIPSE/OPM Flow.
```

File: docs/reference/configuration/keywords.rst (hunk 3 above)
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:137]:
```suggestion
        copied to the runpath folder. Note that support for parsing the ECLIPSE/OPM Flow
```

File: docs/reference/configuration/keywords.rst [hunk 7]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
...
-        entry number 5 otherwise plus one cpu for the master node.
+        ert will search for ``PARALLEL`` in the data file, and if not found, will
+        search for ``SLAVES`` and update :ref:`NUM_CPU <num_cpu>` according to how
+        many nodes are found.
+        The entry number five of the keyword ``SLAVES`` determines the number of processors
+        to run the slave on.
+        Ert does not parse the data files of the nodes, and assumes one CPU per node
+        where the fifth entry is not set.
+        The number of CPUs available for the master node is the number specified in the
+        fifth entry of ``SLAVES`` plus one.
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:151]: I do not think it is true - I read the documentation before as there will be one CPU node for master and whatever specified for slaves

Comment (kvashchuka) [reply docs/reference/configuration/keywords.rst:151]: I will double-check it now

Comment (dafeda) [reply docs/reference/configuration/keywords.rst:151]: Not sure I understand. Do you mean that you read the previous version of the documentation differently than what I suggest here?

Comment (kvashchuka) [reply docs/reference/configuration/keywords.rst:151]: yes

Comment (dafeda) [reply docs/reference/configuration/keywords.rst:151]: This is the previous version:

"... and the number of entry number 5 otherwise plus one cpu for the master node."
 
 I find this really hard to parse.
 What does the `plus` here refer to you think?

Comment (kvashchuka) [reply docs/reference/configuration/keywords.rst:151]: So after a brief discussion with @berland we agreed on the fact the the total number of CPUs needed is the sum of whatever is specified in the 5th item of each slave reservoir (with default 1) plus one CPU for the master process. I can come up with suggestion that reflects that

File: tests/unit_tests/test_run_path_creation.py (hunk 4 above)
Comment (oyvindeide) [reply tests/unit_tests/test_run_path_creation.py:208]: Ahh, I must have missed it, sorry.

File: docs/reference/configuration/keywords.rst [hunk 8]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
...
-        data file, see  :ref:`RUN_TEMPLATE <run_template>`.
+    Example:
 
+    .. code-block::
 
+        -- Load the data file called ECLIPSE.DATA
+        DATA_FILE ECLIPSE.DATA
 
+    See the :ref:`DATA_KW <data_kw>` keyword which can be used to utilize more template
+    functionality in the Eclipse/flow datafile.
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:161]:
```suggestion
    functionality in the ECLIPSE/OPM Flow datafile.
```

File: tests/unit_tests/test_run_path_creation.py [hunk 9]
```diff
@@ -194,6 +194,33 @@ def test_run_template_replace_in_ecl_data_file(key, expected, prior_ensemble):
...
 
+def test_that_error_is_raised_when_data_file_is_badly_encoded(prior_ensemble):
+    config_text = dedent(
+        """
+    NUM_REALIZATIONS 1
+    ECLBASE ECL_CASE<IENS>
+    DATA_FILE MY_DATA_FILE.DATA
+    """
+    )
+    Path("MY_DATA_FILE.DATA").write_text("I WANT TO REPLACE:<DATE>", encoding="utf-8")
```
Comment (oyvindeide) [inline tests/unit_tests/test_run_path_creation.py:205]: This does not happen in a tmpdir, so puts file in the source directory.

File: tests/unit_tests/test_run_path_creation.py (hunk 4 above)
Comment (oyvindeide) [reply tests/unit_tests/test_run_path_creation.py:208]: Come to think of it, why does not your test raise the config error when you create `ErtConfig`?

File: tests/unit_tests/test_run_path_creation.py (hunk 9 above)
Comment (dafeda) [reply tests/unit_tests/test_run_path_creation.py:205]: Nice catch. Fixed.

File: docs/reference/configuration/keywords.rst [hunk 10]
```diff
@@ -128,44 +128,37 @@ Commonly used keywords
...
-        entry number 5 otherwise plus one cpu for the master node.
+        ert will search for ``PARALLEL`` in the data file, and if not found, will
+        search for ``SLAVES`` and update :ref:`NUM_CPU <num_cpu>` according to how
+        many nodes are found.
+        The entry number five of the keyword ``SLAVES`` determines the number of processors
+        to run the slave reservoir on.
+        Ert does not parse the data files of the nodes, and assumes one CPU per node
+        where the fifth entry is not set.
+        The number of CPUs available for the master node is the number specified in the
+        fifth entry of ``SLAVES`` plus one.
```
Comment (kvashchuka) [inline docs/reference/configuration/keywords.rst:151]:
```suggestion
                ert will search for ``PARALLEL`` in the data file, and if not found, will search for ``SLAVES`` and update :ref:`NUM_CPU <num_cpu>` according to how
        many nodes are found.
        The fifth item of the keyword ``SLAVES`` determines the number of processors
        to run the slave reservoir on with the default value of `1` when the fifth item is not set.
        Note that ert does not parse the data files directly and relies on what is specified in the ``SLAVES`` keyword.
        Therefore, the total number of CPUs is the sum of the number of processes for each slave reservoir plus one for the master process.
```

Comment (dafeda) [reply docs/reference/configuration/keywords.rst:151]: Real nice 👍

**Review by oyvindeide — APPROVED** [review]
LGTM!
